ERC721/serverless-NFT/1-07.md

```diff
@@ -0,0 +1,138 @@
+## Step 7: Make Migration for Updates
+
+If you change your schema, or endpoints, or the `tokenURI` standard changes, you need to update your migration contract. Let's make a migration file that we can run for that need.
+
+Inside _Token.sol_ import the _Ownable.sol_ contract from open-zeppelin, inherit it in your token, and define your contract as `Ownable`. Then add a function that can update the `metadata` contract address and restrict the access with the `onlyOwner` modifier. We're referring to `metadata` as the type `Metadata` but for all practical purposes this is just an `address`.
+
+```solidity
+pragma solidity ^0.5.0;
+import "openzeppelin-solidity/contracts/token/ERC721/ERC721Full.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "./Metadata.sol";
+
+/**
+ * The Token contract does this and that...
+ */
+contract Token is ERC721Full, Ownable {
+    Metadata metadata;
+    constructor(string memory name, string memory symbol, Metadata _metadata) public
+        ERC721Full(name, symbol)
+    {
+        metadata = _metadata;
+    }
+
+    function updateMetadata(Metadata _metadata) public onlyOwner {
+        metadata = _metadata;
+    }
+
+    function tokenURI(uint _tokenId) external view returns (string memory _infoUrl) {
+        return metadata.tokenURI(_tokenId);
+    }
+
+}
+```
+
+Duplicate the file  _2_deploy_contracts.js_ and call it  _3_update_metadata.js_:
+
+```bash
+cp ./migrations/2_deploy_contracts.js  ./migrations/3_update_metadata.js
+```
+
+Change the metadata `deploy` in _3_update_metadata.js_ so that it contains an object that specified this contract will be replaced:
+
+```javascript
+await deployer.deploy(Metadata, {replace: true})
+```
+
+Then remove the deployment of the token, and make the already deployed token findable:
+
+```javascript
+// Deployed Token.sol
+// await deployer.deploy(Token, 'Token Name', 'Token Symbol', metadata.address)
+let token = await Token.deployed()
+console.log(_ + 'Token deployed at: ' + token.address)
+```
+
+Then update the token with the new metadata address
+
+```javascript
+await token.updateMetadata(metadata.address)
+console.log(_ + 'Token metadata updated to ' + metadata.address)
+```
+
+<!-- TODO: What does nonces get weird mean? -->
+
+Run the migration to make sure it worked. If the nonces get weird try deleting the _build_ directory.
```

Comment:
@okwme Is there a better way to describe what

>  If the nonces get weird

means?

Reply:
```suggestion
Run the migration to make sure it worked. If you're getting an error about the nonces being incorrect this might have to do with an out of sync truffle build. Try deleting the _build_ directory and running it again.
```